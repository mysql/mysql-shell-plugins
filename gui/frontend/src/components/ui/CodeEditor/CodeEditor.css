/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

.msg.codeEditor {
    /* Restore selection ability. Very important for Safari. */
    user-select: text !important;
    --loading-progress-color: gray;
    --waiting-progress-color: gray;
}

.msg.editorHost {
    flex: 1 1 auto;
    overflow: hidden;
}

.msg.codeEditor .zoneHost {
    z-index: 2;
    padding-right: 16px;
}

.msg.codeEditor .textHost {
    white-space: pre-wrap;
}

.msg.codeEditor .zoneHost > .renderTarget {
    background-color: var(--editorZone-background);
    color: var(--editorZone-foreground);

    overflow: auto;
    display: flex;
    flex-direction: column;
    height: 100%;

    padding: 4px 4px 0 0;
}

.msg.codeEditor .zoneHost > .renderTarget > .resultText {
    line-height: 22px;
    height: 100%;
}

.msg.codeEditor .zoneMargin.focused::before {
    background-color: white;
}

.msg.codeEditor .view-line .error {
    background: url(../../../assets/images/squiggle.svg) bottom left repeat-x;
}

.msg.codeEditor .editorPromptFirst.loading::after {
    content: "";
    display: block;
    width: 60%;
    height: 3px;

    position: absolute;
    bottom: 0;

    background-image: repeating-linear-gradient(-45deg,
            var(--loading-progress-color) 25%,
            transparent 25%,
            transparent 50%,
            var(--loading-progress-color) 50%,
            var(--loading-progress-color) 75%,
            transparent 75%,
            transparent);
    background-size: 7px 3px;
    background-position: 0;

    animation: linearMovement 1s;
    animation-iteration-count: infinite;
}

@keyframes linearMovement {
    0% {
        background-position: 2px;
    }

    100% {
        background-position: 9px;
    }
}

.msg.codeEditor .editorPromptFirst.waiting::after {
    content: "";
    display: block;
    width: 60%;
    height: 3px;

    position: absolute;
    bottom: 0;

    background: linear-gradient(to right,
            var(--waiting-progress-color) 50%,
            transparent 50%,
            transparent);
    background-position: 0;
    background-repeat: no-repeat;

    animation: bouncingMovement 1s;
    animation-iteration-count: infinite;
    animation-direction: alternate;
    animation-timing-function: linear;
}

@keyframes bouncingMovement {
    0% {
        background-position: -10px;
    }

    100% {
        background-position: 22px;
    }
}

.msg.codeEditor .zoneMargin::before {
    content: "";
    display: block;
    position: absolute;

    width: 10px;
    height: inherit;
    right: 42px;
}

.msg.codeEditor .editorPromptFirst.typescript::before {
    content: "ts>";
    display: block;

    margin-left: 13px;

    font-size: 0.9rem;

    color: var(--editorPrompt-primary-ts);
}

.msg.codeEditor.decorationSet-standard .editorPrompt.typescript::before {
    content: "";
    display: block;
    position: absolute;

    width: 7px;
    right: 10px;
    top: -8px;
    height: 52px;

    mask-image: url(../../../assets/images/editor-side-line.svg);
    background-color: var(--editorPrompt-secondary-ts);
    mask-size: 5px;
    mask-repeat: no-repeat;
}

.msg.codeEditor.decorationSet-alternative .editorPrompt.typescript::before {
    content: "";
    display: block;
    opacity: 0.5;
    position: absolute;

    width: 10px;
    height: inherit;
    right: 12px;

    background-image: repeating-linear-gradient(-45deg,
            var(--editorPrompt-secondary-ts) 7%,
            transparent 7%,
            transparent 50%,
            var(--editorPrompt-secondary-ts) 50%,
            var(--editorPrompt-secondary-ts) 57%,
            transparent 57%);
    background-size: 6px 6px;
    background-position: 0;
    background-attachment: fixed;
}

.msg.codeEditor .editorPromptFirst.javascript::before {
    content: "js>";
    display: block;

    margin-left: 13px;

    font-size: 0.9rem;

    color: var(--editorPrompt-primary-js);
}

.msg.codeEditor.decorationSet-standard .editorPrompt.javascript::before {
    content: "";
    display: block;
    position: absolute;

    width: 7px;
    right: 10px;
    top: -8px;
    height: 52px;

    mask-image: url(../../../assets/images/editor-side-line.svg);
    background-color: var(--editorPrompt-secondary-js);
    mask-size: 5px;
    mask-repeat: no-repeat;
}

.msg.codeEditor.decorationSet-alternative .editorPrompt.javascript::before {
    content: "";
    display: block;
    opacity: 0.5;
    position: absolute;

    width: 10px;
    height: inherit;
    right: 12px;

    background-image: repeating-linear-gradient(-45deg,
            var(--editorPrompt-secondary-js) 7%,
            transparent 7%,
            transparent 50%,
            var(--editorPrompt-secondary-js) 50%,
            var(--editorPrompt-secondary-js) 57%,
            transparent 57%);
    background-size: 6px 6px;
    background-position: 0;
    background-attachment: fixed;
}

.msg.codeEditor .editorPromptFirst.sql::before,
.msg.codeEditor .editorPromptFirst.mysql::before {
    content: "sql>";
    display: block;
    margin-left: 4px;
    font-size: 0.9rem;
    color: var(--editorPrompt-primary-sql);
}

.msg.codeEditor.decorationSet-standard .editorPrompt.sql::before,
.msg.codeEditor.decorationSet-standard .editorPrompt.mysql::before {
    content: "";
    display: block;
    position: absolute;

    width: 7px;
    right: 10px;
    top: -8px;
    height: 52px;

    mask-image: url(../../../assets/images/editor-side-line.svg);
    background-color: var(--editorPrompt-secondary-sql);
    mask-size: 5px;
    mask-repeat: no-repeat;
}

.msg.codeEditor.decorationSet-alternative .editorPrompt.sql::before,
.msg.codeEditor.decorationSet-alternative .editorPrompt.mysql::before {
    content: "";
    display: block;
    opacity: 0.5;
    position: absolute;

    width: 10px;
    height: inherit;
    right: 12px;

    background-image: repeating-linear-gradient(-45deg,
            var(--editorPrompt-secondary-sql) 7%,
            transparent 7%,
            transparent 50%,
            var(--editorPrompt-secondary-sql) 50%,
            var(--editorPrompt-secondary-sql) 57%,
            transparent 57%);
    background-size: 6px 6px;
    background-position: 0;
    background-attachment: fixed;
}

.msg.codeEditor .editorPromptFirst.python::before {
    content: "py>";
    display: block;
    margin-left: 13px;
    font-size: 0.9rem;
    color: var(--editorPrompt-primary-py);
}

.msg.codeEditor.decorationSet-standard .editorPrompt.python::before {
    content: "";
    display: block;
    position: absolute;

    width: 7px;
    right: 10px;
    top: -8px;
    height: 52px;

    mask-image: url(../../../assets/images/editor-side-line.svg);
    background-color: var(--editorPrompt-secondary-py);
    mask-size: 5px;
    mask-repeat: no-repeat;
}

.msg.codeEditor.decorationSet-alternative .editorPrompt.python::before {
    content: "";
    display: block;
    opacity: 0.5;
    position: absolute;

    width: 10px;
    height: inherit;
    right: 12px;

    background-image: repeating-linear-gradient(-45deg,
            var(--editorPrompt-secondary-py) 7%,
            transparent 7%,
            transparent 50%,
            var(--editorPrompt-secondary-py) 50%,
            var(--editorPrompt-secondary-py) 57%,
            transparent 57%);
    background-size: 6px 6px;
    background-position: 0;
    background-attachment: fixed;
}

.msg.codeEditor .statementStart::after {
    content: "";
    display: block;
    position: absolute;

    border-radius: 50%;
    width: 4px;
    height: 4px;

    right: 4px;
    top: 50%;
    background: #0096FF;
}

.msg.codeEditor .margin-view-overlays .codicon::before {
    /* Make the code folding icons left aligned, instead of centered. */
    margin-right: auto;
}
