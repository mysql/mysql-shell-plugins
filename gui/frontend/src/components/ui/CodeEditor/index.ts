/*
 * Copyright (c) 2020, 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { languages, editor, IPosition } from "monaco-editor/esm/vs/editor/editor.api.js";
import { IStatementSpan } from "../../../parsing/parser-common.js";

import { IContextProvider, IExecuteResultReference } from "../../../script-execution/index.js";
import { EditorLanguage } from "../../../supplement/index.js";

export type {
    IMarkdownString, IRange, IPosition, IDisposable, CancellationToken, Thenable,
} from "monaco-editor/esm/vs/editor/editor.api.js";

export {
    editor as Monaco, languages, KeyMod, KeyCode, Selection, Range, Position, Uri,
} from "monaco-editor/esm/vs/editor/editor.api.js";

export type ProviderResult<T> = languages.ProviderResult<T>;

export type CompletionList = languages.CompletionList;
export type CompletionItem = languages.CompletionItem;
export type CompletionItemLabel = languages.CompletionItemLabel;
export type CompletionContext = languages.CompletionContext;

export type Hover = languages.Hover;

export type SignatureHelpContext = languages.SignatureHelpContext;
export type SignatureHelpResult = languages.SignatureHelpResult;
export type SignatureHelp = languages.SignatureHelp;
export type SignatureInformation = languages.SignatureInformation;
export type ParameterInformation = languages.ParameterInformation;

export type DocumentHighlight = languages.DocumentHighlight;

export type DocumentSymbol = languages.DocumentSymbol;

export type Definition = languages.Definition;
export type LocationLink = languages.LocationLink;

export type ReferenceContext = languages.ReferenceContext;
export type Location = languages.Location;

export type SelectionRange = languages.SelectionRange;

export type FormattingOptions = languages.FormattingOptions;
export type TextEdit = languages.TextEdit;
export type IWorkspaceTextEdit = languages.IWorkspaceTextEdit;

export type FoldingRange = languages.FoldingRange;
export type FoldingContext = languages.FoldingContext;

export type WorkspaceEdit = languages.WorkspaceEdit;
export type RenameLocation = languages.RenameLocation;
export type Rejection = languages.Rejection;

export type TypeScriptWorker = languages.typescript.TypeScriptWorker;

export type TokensProvider = languages.TokensProvider;
export type DocumentSemanticTokensProvider = languages.DocumentSemanticTokensProvider;
export type SemanticTokens = languages.SemanticTokens;
export type SemanticTokensLegend = languages.SemanticTokensLegend;

export type ICodeEditorViewState = editor.ICodeEditorViewState;

export interface IExecutionContextState {
    /** The start line for this context. */
    start: number;

    /** The end line for this context. */
    end: number;

    /** The language used in the context. */
    language: EditorLanguage;

    /** Optionally: an attached result (reference). */
    result?: IExecuteResultReference;

    /** The height of the result pane. Only considered if the pane is not maximized. */
    currentHeight?: number;

    /** The index of the active set, in a multi set result. */
    currentSet?: number;

    /** When true no editor is shown and the result pane takes the entire space. */
    maximizeResultPane?: boolean;

    /** Stored statement ranges. */
    statements: IStatementSpan[];
}

/** Options that control how SQL statements are executed. */
export interface IScriptExecutionOptions {
    /** Any additional named parameters for placeholders in the query. */
    params?: Array<[string, string]>;

    /**
     * Used when executing SQL statements to tell the executor to add a hint to SELECT statements to use the secondary
     * engine (usually HeatWave).
     */
    forceSecondaryEngine?: boolean;

    /**
     * Determines where the SQL code comes from that must be executed. If that's a position it means
     * to run only the code at that (caret) position. If not given at all run the statements touched by the editor
     * selection or all statements, if there's no selection.
     */
    source?: IPosition;

    /** When true render the query and the result as plain text. */
    asText?: boolean;
}

/** Determines the behavior of the code editor. */
export enum CodeEditorMode {
    Standard,

    /** Different set of code completion values. Allows for language switch commands. */
    Terminal
}

/** A stripped down model interface for the various code editor providers. */
export interface IProviderEditorModel extends editor.ITextModel {
    executionContexts?: IContextProvider;

    /** Functionality differs depending on where the code editor is used. */
    editorMode: CodeEditorMode;
}

/** The structure of a monaco-editor language definition. */
export interface ILanguageDefinition {
    language: languages.IMonarchLanguage;
    languageConfiguration: languages.LanguageConfiguration;
}

/** Base scope names used for syntax highlighting. These are combined with the modifiers below. */
export const tokenTypes = [
    "keyword",
    "identifier",
    "operator",
    "delimiter",
    "markup",
    "number",
    "string",
    "comment",
    "regexp",
    "support",
    "entity",
    "type",
    "operator",
    "namespace",
    "struct",
    "class",
    "interface",
    "enum",
    "typeParameter",
    "function",
    "member",
    "macro",
    "variable",
    "parameter",
    "property",
    "label",
    "command",
];

/**
 * Identifiers attached to a base scope to form a full scope. Any number of modifiers can be used, separated by a dot.
 *
 * Important note: each entry is represented by a bit in the final unsigned integer. This means that the maximum number
 * of entries is 32.
 */
export const tokenModifiers = [
    "other",
    "bracket",
    "square",
    "parenthesis",
    "float",
    "hex",
    "octal",
    "binary",
    "regexp",
    "escape",
    "invalid",
    "control",
    "line",
    "block",
    "doc",
    "quoted",
    "single",
    "double",
    "function",
    "variable",
    "user",
    "system",
    "entity",
    "name",
    "identifier",
    "language",
    "predefined",
    "angle",
];
