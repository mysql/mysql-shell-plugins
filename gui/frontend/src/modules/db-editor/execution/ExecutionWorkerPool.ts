/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { WorkerPool } from "../../../supplement/WorkerPool";
import { IConsoleWorkerTaskData, IConsoleWorkerResultData } from "../console.worker-types";

/* eslint import/no-webpack-loader-syntax: off */
import ConsoleWorker from "worker-loader?filename=static/workers/[name].[contenthash].js!./console.worker";

// A specialized worker pool for interactive console tasks.
export class ExecutionWorkerPool extends WorkerPool<IConsoleWorkerTaskData, IConsoleWorkerResultData> {

    private static nextId = 0;

    protected createNewWorker(): ConsoleWorker {
        const worker = new ConsoleWorker();
        worker.id = "ew" + String(ExecutionWorkerPool.nextId++);

        return worker;
    }
}
