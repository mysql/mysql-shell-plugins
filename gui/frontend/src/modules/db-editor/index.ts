/*
 * Copyright (c) 2020, 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

/// <reference types="./assets/typings/scripting-runtime.d.ts" />

import { ComponentChild, createContext } from "preact";

import type { ConnectionDataModel, ConnectionDataModelEntry } from "../../data-models/ConnectionDataModel.js";
import type { OciDataModel, OciDataModelEntry } from "../../data-models/OciDataModel.js";
import type { OpenDocumentDataModel, OpenDocumentDataModelEntry } from "../../data-models/OpenDocumentDataModel.js";
import type { ShellTaskDataModel } from "../../data-models/ShellTaskDataModel.js";

/** This is the base structure for a tree in the sidebar UI. */
export interface IBaseTreeItem<T> {
    /** This is a copy of the id from the data model entry to allow searching for it. */
    id: string;

    /** The data model entry that this interface represents. */
    dataModelEntry: T;

    /** The caption for the UI. Can change at any time. */
    caption: string;

    children?: Array<IBaseTreeItem<T>>;
}

/** Comprises the parts of a fully qualified name. */
export type QualifiedName = {
    /** Not assigned for connections. */
    schema?: string;

    /** Not assigned for connections and schema entries. Holds a table or view name. */
    table?: string;

    /** Assigned only for entries that have either a table or a schema as parent. */
    name?: string;
};

/** Represents the visual expression of a connection data model entry. */
export interface IConnectionTreeItem extends IBaseTreeItem<ConnectionDataModelEntry> {
    /** The parts of the FQN to access a db object. */
    qualifiedName: QualifiedName;

    children?: IConnectionTreeItem[];
}

/** Represents the visual expression of a connection data model entry. */
export interface IDocumentTreeItem extends IBaseTreeItem<OpenDocumentDataModelEntry> {
    children?: IDocumentTreeItem[];
}

/** Represents the visual expression of an OCI data model entry. */
export interface IOciTreeItem extends IBaseTreeItem<OciDataModelEntry> {
    children?: IOciTreeItem[];
}

/**
 * Lists of items to be added to a toolbar. The receiving page can use this as base to add it's own
 * items, before it actually renders the toolbar.
 */
export interface IToolbarItems {
    /** Items used to navigate between documents and to create new documents. */
    navigation: ComponentChild[];

    /** Items to execute blocks, including transaction toggles. */
    execution: ComponentChild[];

    /** Items related to the code editor (format, clean, word wrap etc.) */
    editor: ComponentChild[];

    /** Additional items for special functionality, which are right aligned in the toolbar. */
    auxillary: ComponentChild[];
}

/** Predefined color schemes in graphs. */
export type ColorScheme = "classic" | "delectable" | "trello" | "brewing" | "light" | "lively" | "grays";

/** Cached data for the performance dashboard. */
export interface ISavedGraphData {
    /** The currently selected graph color scheme (default: classic). */
    activeColorScheme: ColorScheme;

    /** The currently selected display interval in line graphs. */
    displayInterval: number;

    /** The timestamp when this snapshot was taken. */
    timestamp: number;

    /** Last read values from the server (for computing differences). */
    currentValues: Map<string, number>;

    /** Computed values for labels and pie graphs.  */
    computedValues: { [key: string]: number; };

    /** Stored data per line graph. */
    series: Map<string, IXYDatum[]>;
}

/** The preact context holding shared data for all pages in the DB editor module. */
export type DBEditorContextType = {
    connectionsDataModel: ConnectionDataModel;
    documentDataModel: OpenDocumentDataModel;
    ociDataModel: OciDataModel;
    shellTaskDataModel: ShellTaskDataModel;
};

/**
 * The context class for the DB editor module. This is used to share data between the various
 * components in the module.
 * Note: have to include `undefined` in the type list here, to avoid being forced to provide a
 *       default value in the `createContext` call.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export const DBEditorContext = createContext<DBEditorContextType | undefined>(undefined);

/** Describes possible results for a sidebar command. */
export interface ISideBarCommandResult {
    /** Indicates whether the command was successfully executed or cancelled (by the user or an error). */
    success: boolean;

    /** For MRS related commands: indicates the MRS service that was modified. */
    mrsServiceId?: string;

    /** For MRS related commands: indicates the MRS schema that was modified. */
    mrsSchemaId?: string;
}
