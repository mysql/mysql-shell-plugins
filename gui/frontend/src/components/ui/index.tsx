/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

import { ICodeEditorOptions } from "./CodeEditor";

export const defaultEditorOptions: ICodeEditorOptions = {
    tabSize: 4,
    indentSize: 4,
    insertSpaces: true,

    defaultEOL: "LF",
    trimAutoWhitespace: true,
};

export * from "./Component/Component";
export * from "./Container/Container";

export * from "./ActivityBar/ActivityBar";
export * from "./ActivityBar/ActivityBarItem";

export * from "./Button/Button";
export * from "./Checkbox/Checkbox";
export * from "./Radiobutton/Radiobutton";
export * from "./Dialog/Dialog";
export * from "./Icon/Icon";
export * from "./Codicon";
export * from "./Image/Image";
export * from "./Popup/Popup";
export * from "./Portal/Portal";

export * from "./Grid/Grid";
export * from "./Grid/GridCell";

export * from "./Label/Label";
export * from "./Input/Input";
export * from "./Search/Search";
export * from "./Message/Message";

export * from "./SplitContainer/SplitContainer";

export * from "./Accordion/Accordion";
export * from "./Accordion/AccordionItem";

export * from "./Tabview/Tabview";
export * from "./Toggle/Toggle";
export * from "./Divider/Divider";

export * from "./List/List";
export * from "./List/DynamicList";

export * from "./Selector/Selector";
export * from "./Selector/SelectorItem";

export * from "./Menu/Menu";
export * from "./Menu/MenuBar";
export * from "./Menu/MenuItem";

export * from "./Dropdown/Dropdown";

export * from "./Breadcrumb/Breadcrumb";
export * from "./Statusbar/Statusbar";

export * from "./TreeGrid/TreeGrid";

export * from "./ColorPicker/ColorPopup";
export * from "./ColorPicker/ColorField";

export * from "./Slider/Slider";
export * from "./Toolbar/Toolbar";
export * from "./TagInput/TagInput";
export * from "./UpDown/UpDown";
export * from "./Group/Group";
export * from "./Calendar/Calendar";

export * from "./FrontPage/FrontPage";
export * from "./MessageToast/MessageToast";
export * from "./FileSelector/FileSelector";
export * from "./AboutBox/AboutBox";
export * from "./ProgressIndicator/ProgressIndicator";

export * from "./ConnectionTile/ConnectionTile";
export * from "./SessionTile/SessionTile";
export * from "./BrowserTile/BrowserTile";

export * from "./StatusMark/StatusMark";

/**
 * Creates a synthetic event from a native DOM event, for interaction of DOM events with React components.
 *
 * @param event The DOM event to use.
 *
 * @returns A new synthetic event instance.
 */
export const createSyntheticEvent = <T extends Element, E extends Event>(event: E): React.SyntheticEvent<T, E> => {
    let isDefaultPrevented = false;
    let isPropagationStopped = false;
    const preventDefault = (): void => {
        isDefaultPrevented = true;
        event.preventDefault();
    };
    const stopPropagation = (): void => {
        isPropagationStopped = true;
        event.stopPropagation();
    };

    return {
        nativeEvent: event,
        currentTarget: event.currentTarget as EventTarget & T,
        target: event.target as EventTarget & T,
        bubbles: event.bubbles,
        cancelable: event.cancelable,
        defaultPrevented: event.defaultPrevented,
        eventPhase: event.eventPhase,
        isTrusted: event.isTrusted,
        preventDefault,
        stopPropagation,
        timeStamp: event.timeStamp,
        type: event.type,

        isDefaultPrevented: (): boolean => { return isDefaultPrevented; },
        isPropagationStopped: (): boolean => { return isPropagationStopped; },
        persist: (): void => { /**/ },
    };
};
