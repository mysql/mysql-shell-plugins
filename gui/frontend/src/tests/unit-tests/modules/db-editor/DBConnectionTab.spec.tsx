/*
 * Copyright (c) 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mount } from "enzyme";
import React from "react";

import { setupShellForTests } from "../../test-helpers";
import {
    DBConnectionTab, IOpenEditorState, IDBConnectionTabPersistentState,
} from "../../../../modules/db-editor/DBConnectionTab";
import { ShellInterfaceSqlEditor } from "../../../../supplement/ShellInterface/ShellInterfaceSqlEditor";
import { ExecutionWorkerPool } from "../../../../modules/db-editor/execution/ExecutionWorkerPool";
import { DBType } from "../../../../supplement/ShellInterface";
import {
    CodeEditorMode, ICodeEditorModel, IEditorPersistentState,
} from "../../../../components/ui/CodeEditor/CodeEditor";
import { EntityType } from "../../../../modules/db-editor";
import { ExecutionContexts } from "../../../../script-execution/ExecutionContexts";
import { Monaco } from "../../../../components/ui/CodeEditor";
import { MySQLShellLauncher } from "../../../../utilities/MySQLShellLauncher";

jest.mock("../../../../components/ui/CodeEditor/CodeEditor");

describe("DBConnectionTab tests", (): void => {

    let content = `\nprint("typescript");\n\\js\n`;
    content += `\nprint("javascript");\n\\sql\n`;
    content += `\nselect "(my)sql" from dual;\n\\py\n`;
    content += `\nprint("python");\n`;
    const model = Monaco.createModel("", "msg") as ICodeEditorModel;
    model.executionContexts = new ExecutionContexts(undefined, 80024, "", "");
    model.editorMode = CodeEditorMode.Standard;
    model.setValue(content);

    let launcher: MySQLShellLauncher;

    beforeAll(async () => {
        launcher = await setupShellForTests("DBEditor", false, true, "DEBUG2");
    });

    afterAll(async () => {
        await launcher.exitProcess();
    });

    it("Test DBConnectionTab instantiation", () => {
        const wp = new ExecutionWorkerPool();
        const eps: IEditorPersistentState = {
            viewState: null,
            model,
            options: {},
        };

        const edState: IOpenEditorState = {
            state: eps,
            currentVersion: 0,
            type: EntityType.Notebook,
            id: "SQLEditor",
            caption: "",
        };

        const backend = new ShellInterfaceSqlEditor();
        const savedState: IDBConnectionTabPersistentState = {
            backend,
            serverVersion: 1,
            serverEdition: "gpl",
            sqlMode: "mode1",
            editors: [edState],
            scripts: [],
            schemaTree: [],
            explorerState: new Map([["state1", {}]]),
            activeEntry: "SQLEditor",
            currentSchema: "myDb",
            explorerWidth: 200,
            graphData: {
                timestamp: new Date().getTime(),
                activeColorScheme: "grays",
                displayInterval: 300,
                currentValues: new Map(),
                computedValues: {},
                series: new Map(),
            },
        };

        const component = mount<DBConnectionTab>(
            <DBConnectionTab
                connectionId={0}
                dbType={DBType.MySQL}
                savedState={savedState}
                workerPool={wp}
                showAbout={false}
            />,
        );

        const props = component.props();
        expect(props.connectionId).toEqual(0);
        expect(props.dbType).toEqual(DBType.MySQL);
        expect(props.savedState).toEqual(savedState);
        expect(props.workerPool).toEqual(wp);

        component.unmount();
    });

});
