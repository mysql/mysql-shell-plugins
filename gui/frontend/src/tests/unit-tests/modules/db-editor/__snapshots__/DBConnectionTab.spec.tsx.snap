// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DBConnectionTab tests Test DBConnectionTab snapshot 1`] = `<Consumer />`;
